src/Compiler/Checking/Expressions/CheckExpressions.fs

```diff
@@ -7881,39 +8064,372 @@ and TcAnonRecdExpr cenv (overallTy: TType) env tpenv (isStruct, optOrigSynExpr,
 and TcNewAnonRecdExpr cenv (overallTy: TType) env tpenv (isStruct, unsortedFieldIdsAndSynExprsGiven, mWholeExpr) =
 
     let g = cenv.g
-    let unsortedFieldSynExprsGiven = unsortedFieldIdsAndSynExprsGiven |> List.map (fun (_, _, fieldExpr) -> fieldExpr)
-    let unsortedFieldIds = unsortedFieldIdsAndSynExprsGiven |> List.map (fun (synLongIdent, _, _) -> synLongIdent.LongIdent[0]) |> List.toArray
-    let anonInfo, sortedFieldTys = UnifyAnonRecdTypeAndInferCharacteristics env.eContextInfo cenv env.DisplayEnv mWholeExpr overallTy isStruct unsortedFieldIds
-
-    if unsortedFieldIds.Length > 1 then
-        CheckAnonRecdExprDuplicateFields unsortedFieldIds
-
-    // Sort into canonical order
-    let sortedIndexedArgs =
-        unsortedFieldIdsAndSynExprsGiven
-        |> List.indexed
-        |> List.sortBy (fun (i,_) -> unsortedFieldIds[i].idText)
+    let ad = env.eAccessRights
 
-    // Map from sorted indexes to unsorted indexes
-    let sigma = sortedIndexedArgs |> List.map fst |> List.toArray
-    let sortedFieldExprs = sortedIndexedArgs |> List.map snd
+    let maybeAnonRecdTargetTy = tryDestAnonRecdTy g overallTy
+
+    let spreadSrcs, unsortedCheckedFields, anonInfo, tpenv =
+        let (|LeftwardExplicit|NoLeftwardExplicit|) hasLeftwardExplicit = if hasLeftwardExplicit then LeftwardExplicit else NoLeftwardExplicit
+        let LeftwardExplicit = true
+        let NoLeftwardExplicit = false
+
+        let allSortedFieldTys, dedupedSortedFieldTys, tcFieldExprAt =
+            let possibleTargetTyAt =
+                match maybeAnonRecdTargetTy with
+                | ValueSome (anonInfo, tys) ->
+                    let names = anonInfo.SortedNames
+                    let tys = List.toArray tys
+                    fun name ->
+                        let i = Array.BinarySearch (names, name)
+                        if i < 0 then ValueNone
+                        else ValueSome tys[i]
+                | ValueNone -> fun _ -> ValueNone
+
+            // ----------------------------------------------------------------------------------
+            // ----------------------------------------------------------------------------------
+            // TODO: Collect spread src exprs so we don't re-typecheck them?
+            // TODO: It would be better if we could _try_ to typecheck fields and undo if they
```

Comment:
Is there a scenario, where this causes troubles?
Even in case of duplicates, they get a new inference type variable allocated during type checking, I did not find a source of possible clashes in type inference.